Elasticsearch fix search action

r7-mgorny opened the pull request:
## Proposed Changes
MC-677
Fix search example input in help.md
Add exception message in search action

### Description

Describe the proposed changes:

  -

## PR Requirements

Developers, verify you have completed the following items by checking them off:

### Testing

#### Unit Tests

Review our documentation on [generating](https://docs.rapid7.com/insightconnect/unit-test-generation) and [writing](https://docs.rapid7.com/insightconnect/unit-test-primer) plugin unit tests

- [ ] Unit tests written for any new or updated code

#### In-Product Tests

If you are an InsightConnect customer or have access to an InsightConnect instance, the following in-product tests should be done:

- [ ] Screenshot of job output with the plugin changes
- [ ] Screenshot of the changed connection, actions, or triggers input within the InsightConnect workflow builder

### Style

Review the [style guide](https://docs.rapid7.com/insightconnect/style-guide/)

- [ ] For dependencies, pin [OS package](https://docs.rapid7.com/insightconnect/style-guide/#dockerfile) and [Python package](https://docs.rapid7.com/insightconnect/style-guide/#requirements.txt) versions
- [ ] For security, set least privileged account with ``USER nobody`` in the ``Dockerfile`` when possible
- [ ] For size, use the [slim SDK images](https://docs.rapid7.com/insightconnect/sdk-guide/#sdk-guide) when possible: ``komand/python-3-37-slim-plugin`` and ``komand/python-3-37-plugin``
- [ ] For error handling, use of [PluginException](https://docs.rapid7.com/insightconnect/error-handling-in-integrations/#plugin-exceptions) and [ConnectionTestException](https://docs.rapid7.com/insightconnect/error-handling-in-integrations#connection-exceptions)
- [ ] For logging, use [self.logger](https://docs.rapid7.com/insightconnect/sdk-guide/#logging)
- [ ] For docs, use [changelog style](https://docs.rapid7.com/insightconnect/style-guide/#changelog)
- [ ] For docs, validate markdown with ``make validate`` which calls ``mdl`` to lint ``help.md``

### Functional Checklist
- [ ] Work fully completed
- [ ] Functional
  - [ ] Any new actions/triggers include JSON [test files](https://docs.rapid7.com/insightconnect/style-guide/#tests) in the `tests/` directory created with `icon-plugin run -c sample $action > tests/$action.json`
  - [ ] Tests should all pass unless it's a negative test. Negative tests have a naming convention of `tests/$action_bad.json`
  - [ ] Unsuccessful tests should fail by raising an exception causing the plugin to die and an object should be returned on successful test
  - [ ] Add functioning test results to PR, sanitize any output if necessary
    * Single action/trigger `icon-plugin run -T tests/example.json --debug --jq`
    * All actions/triggers shortcut `icon-plugin run -T all --debug --jq` (use PR format at end)
  - [ ] Add functioning run results to PR, sanitize any output if necessary
    * Single action/trigger `icon-plugin run -R tests/example.json --debug --jq`
    * All actions/triggers shortcut `icon-plugin run -R all --debug --jq` (use PR format at end)

## Assessment
### Run

<details>

```
{
  "body": {
    "log": "Connect: Connecting..\nrapid7/https://example.com Step name: search_documents\n",
    "meta": {},
    "output": {
      "hits": {
        "hits": [
          {
            "_id": "eB3N2XoBW9uWLGlBkjnw",
            "_index": "testindexv7",
            "_score": 1,
            "_source": {
              "email": "user@example.com",
              "name": "testnoid"
            },
            "_type": "_doc",
            "_version": 1
          },
          {
            "_id": "DCJG3noBW9uWLGlBw6fT",
            "_index": "testindexv7",
            "_score": 1,
            "_source": {
              "email": "user@example.com",
              "name": "testnoid"
            },
            "_type": "_doc",
            "_version": 1
          },
          {
            "_id": "11",
            "_index": "testindexv7",
            "_score": 1,
            "_source": {
              "email": "user@example.com",
              "name": "test11",
              "number": "123"
            },
            "_type": "_doc",
            "_version": 3
          }
        ],
        "max_score": 1,
        "total": {
          "value": 3
        }
      },
      "shards": {
        "failed": 0,
        "skipped": 0,
        "successful": 1,
        "total": 1
      },
      "timed_out": false,
      "took": 0
    },
    "status": "ok"
  },
  "type": "action_event",
  "version": "v1"
}

```

<summary>
docker run --rm -i rapid7/elasticsearch:3.0.1 --debug run < tests/search_documents_v7_all_matches_query.json
</summary>
</details>

<details>

```
{
  "body": {
    "error": "An error occurred during plugin execution!\n\nWrong input \"query\". Old query input detected. Input shouldn't contain {\"query\": {\"query\": {}}}. Please refer help for more details.",
    "log": "Connect: Connecting..\nrapid7/https://example.com Step name: search_documents\nAn error occurred during plugin execution!\n\nWrong input \"query\". Old query input detected. Input shouldn't contain {\"query\": {\"query\": {}}}. Please refer help for more details.\nTraceback (most recent call last):\n  File \"/usr/local/lib/https://example.com\", line 326, in handle_step\n    output = https://example.com\n  File \"/usr/local/lib/https://example.com\", line 476, in start_step\n    output = func(params)\n  File \"/usr/local/lib/https://example.com\", line 22, in run\n    raise PluginException(\https://example.com An error occurred during plugin execution!\n\nWrong input \"query\". Old query input detected. Input shouldn't contain {\"query\": {\"query\": {}}}. Please refer help for more details.\n",
    "meta": {},
    "status": "error"
  },
  "type": "action_event",
  "version": "v1"
}

```

<summary>
docker run --rm -i rapid7/elasticsearch:3.0.1 --debug run < tests/search_documents_v7_all_matches_query_bad.json
</summary>
</details>

### Validators
<details>

```
[*] Validating plugin with all validators at .

[*] Running Integration Validators...
[*] Executing validator HelpValidator
[*] Executing validator ChangelogValidator
[*] Executing validator CloudReadyConnectionCredentialTokenValidator
[*] Executing validator RequiredKeysValidator
[*] Executing validator UseCaseValidator
[*] Executing validator SpecPropertiesValidator
[*] Executing validator SpecVersionValidator
[*] Executing validator FilesValidator
[*] Executing validator TagValidator
[*] Executing validator DescriptionValidator
[*] Executing validator TitleValidator
[*] Executing validator VendorValidator
[*] Executing validator DefaultValueValidator
[*] Executing validator IconValidator
[*] Executing validator RequiredValidator
[*] Executing validator VersionValidator
[*] Executing validator DockerfileParentValidator
[*] Executing validator ProfanityValidator
[*] Executing validator AcronymValidator
[*] Executing validator JSONValidator
[*] Executing validator OutputValidator
[*] Executing validator RegenerationValidator
[*] Executing validator HelpInputOutputValidator
[*] Executing validator SupportValidator
[*] Executing validator RuntimeValidator
[*] Executing validator VersionPinValidator
[*] Executing validator EncodingValidator
[*] Executing validator ExampleInputValidator
[*] Executing validator CloudReadyValidator
[*] Executing validator UnapprovedKeywordsValidator
[*] Executing validator HelpExampleValidator
[*] Executing validator ExceptionValidator
[*] Executing validator CredentialsValidator
[*] Executing validator PasswordValidator
[*] Executing validator PrintValidator
[*] Executing validator ConfidentialValidator
[*] Executing validator DockerValidator
[*] Executing validator URLValidator
WARNING: URLs found that return a 4xx code. Verify they are publicly accessible and if not, update with a working URL.
violation: plugin.spec.yaml[30]: https://www.example.com:9243
[*] Plugin failed validation! The following validation errors occurred:

Validator "HelpExampleValidator" failed! 
	Cause: Invalid JSON example identified in Example input of action titled Cluster Health. Please correct the JSON example in help.md.
	Invalid JSON example identified in Example input of action titled Input. Please correct the JSON example in help.md.
	Invalid JSON example identified in Example input of action titled Input. Please correct the JSON example in help.md.
	Invalid JSON example identified in Example input of action titled Input. Please correct the JSON example in help.md.
	Invalid JSON example identified in Example input of trigger titled Poll Documents. Please correct the JSON example in help.md.


----
[*] Total time elapsed: 14225.448999999999ms

```

<summary>
icon-validate --all .
</summary>
</details>

<details>

```
[*] Use ``make menu`` for available targets
[*] Including available Makefiles: ../../tools/Makefiles/Colors.mk ../../tools/Makefiles/Helpers.mk
--
[*] Running validators
[*] Validating plugin at .

[*] Running Integration Validators...
[*] Executing validator HelpValidator
[*] Executing validator ChangelogValidator
[*] Executing validator CloudReadyConnectionCredentialTokenValidator
[*] Executing validator RequiredKeysValidator
[*] Executing validator UseCaseValidator
[*] Executing validator SpecPropertiesValidator
[*] Executing validator SpecVersionValidator
[*] Executing validator FilesValidator
[*] Executing validator TagValidator
[*] Executing validator DescriptionValidator
[*] Executing validator TitleValidator
[*] Executing validator VendorValidator
[*] Executing validator DefaultValueValidator
[*] Executing validator IconValidator
[*] Executing validator RequiredValidator
[*] Executing validator VersionValidator
[*] Executing validator DockerfileParentValidator
[*] Executing validator ProfanityValidator
[*] Executing validator AcronymValidator
[*] Executing validator JSONValidator
[*] Executing validator OutputValidator
[*] Executing validator RegenerationValidator
[*] Executing validator HelpInputOutputValidator
[*] Executing validator SupportValidator
[*] Executing validator RuntimeValidator
[*] Executing validator VersionPinValidator
[*] Executing validator EncodingValidator
[*] Executing validator ExampleInputValidator
[*] Executing validator CloudReadyValidator
[*] Executing validator UnapprovedKeywordsValidator
[*] Executing validator HelpExampleValidator
[*] Plugin failed validation! The following validation errors occurred:

Validator "HelpExampleValidator" failed! 
	Cause: Invalid JSON example identified in Example input of action titled Cluster Health. Please correct the JSON example in help.md.
	Invalid JSON example identified in Example input of action titled Input. Please correct the JSON example in help.md.
	Invalid JSON example identified in Example input of action titled Input. Please correct the JSON example in help.md.
	Invalid JSON example identified in Example input of action titled Input. Please correct the JSON example in help.md.
	Invalid JSON example identified in Example input of trigger titled Poll Documents. Please correct the JSON example in help.md.


----
[*] Total time elapsed: 2191.203ms

```

<summary>
make validate
</summary>
</details>

jschipp-r7 commented:
@r7-mgorny The output looks escaped, can we fix that so it's more readable

```
Old query input detected. Input shouldn't contain {\"query\": {\"query\": {}}}.
```

Also, this plugin supports ElasticSearch 6 and 7. Does the old query work in 6? do we need to check the ES version in the if-statement?. We'll want to make sure it will still work for 6 and 7 users.

r7-mgorny commented:
## Assessment
### Run

<details>

```
{
  "body": {
    "log": "Connect: Connecting..\nrapid7/https://example.com Step name: search_documents\n",
    "meta": {},
    "output": {
      "hits": {
        "hits": [
          {
            "_id": "eB3N2XoBW9uWLGlBkjnw",
            "_index": "testindexv7",
            "_score": 1,
            "_source": {
              "email": "https://example.com",
              "name": "testnoid"
            },
            "_type": "_doc"
          },
          {
            "_id": "DCJG3noBW9uWLGlBw6fT",
            "_index": "testindexv7",
            "_score": 1,
            "_source": {
              "email": "https://example.com",
              "name": "testnoid"
            },
            "_type": "_doc"
          },
          {
            "_id": "11",
            "_index": "testindexv7",
            "_score": 1,
            "_source": {
              "email": "https://example.com",
              "name": "test11",
              "number": "123"
            },
            "_type": "_doc"
          }
        ],
        "max_score": 1,
        "total": {
          "value": 3
        }
      },
      "shards": {
        "failed": 0,
        "skipped": 0,
        "successful": 1,
        "total": 1
      },
      "timed_out": false,
      "took": 0
    },
    "status": "ok"
  },
  "type": "action_event",
  "version": "v1"
}

```

<summary>
docker run --rm -i rapid7/elasticsearch:3.0.1 --debug run < tests.back/search_documents_v7_without_query.json
</summary>
</details>

<details>

```
{
  "body": {
    "log": "Connect: Connecting..\nrapid7/https://example.com Step name: search_documents\n",
    "meta": {},
    "output": {
      "hits": {
        "hits": [
          {
            "_id": "eB3N2XoBW9uWLGlBkjnw",
            "_index": "testindexv7",
            "_score": 1,
            "_source": {
              "email": "https://example.com",
              "name": "testnoid"
            },
            "_type": "_doc",
            "_version": 1
          },
          {
            "_id": "DCJG3noBW9uWLGlBw6fT",
            "_index": "testindexv7",
            "_score": 1,
            "_source": {
              "email": "https://example.com",
              "name": "testnoid"
            },
            "_type": "_doc",
            "_version": 1
          },
          {
            "_id": "11",
            "_index": "testindexv7",
            "_score": 1,
            "_source": {
              "email": "https://example.com",
              "name": "test11",
              "number": "123"
            },
            "_type": "_doc",
            "_version": 3
          }
        ],
        "max_score": 1,
        "total": {
          "value": 3
        }
      },
      "shards": {
        "failed": 0,
        "skipped": 0,
        "successful": 1,
        "total": 1
      },
      "timed_out": false,
      "took": 0
    },
    "status": "ok"
  },
  "type": "action_event",
  "version": "v1"
}

```

<summary>
docker run --rm -i rapid7/elasticsearch:3.0.1 --debug run < tests.back/search_documents_v7_all_matches_query.json
</summary>
</details>

<details>

```
{
  "body": {
    "error": "An error occurred during plugin execution!\n\nWrong input 'query'. Old query style detected during input. Input shouldn't contain {'query': {'query': ...}}. Please refer help for more details or elasticsearch API documentation: https://example.com",
    "log": "Connect: Connecting..\nrapid7/https://example.com Step name: search_documents\nAn error occurred during plugin execution!\n\nWrong input 'query'. Old query style detected during input. Input shouldn't contain {'query': {'query': ...}}. Please refer help for more details or elasticsearch API documentation: https://example.com\nTraceback (most recent call last):\n  File \"/usr/local/lib/https://example.com\", line 326, in handle_step\n    output = https://example.com\n  File \"/usr/local/lib/https://example.com\", line 476, in start_step\n    output = func(params)\n  File \"/usr/local/lib/https://example.com\", line 22, in run\n    raise PluginException(\https://example.com An error occurred during plugin execution!\n\nWrong input 'query'. Old query style detected during input. Input shouldn't contain {'query': {'query': ...}}. Please refer help for more details or elasticsearch API documentation: https://example.com\n",
    "meta": {},
    "status": "error"
  },
  "type": "action_event",
  "version": "v1"
}

```

<summary>
docker run --rm -i rapid7/elasticsearch:3.0.1 --debug run < tests.back/search_documents_v7_all_matches_query_bad.json
</summary>
</details>

<details>

```
{
  "body": {
    "log": "Connect: Warning, No Auth Provided\nrapid7/https://example.com Step name: search_documents\n",
    "meta": {},
    "output": {
      "hits": {
        "hits": [
          {
            "_id": "m77N2XoBjtIgvTKiUnVO",
            "_index": "testindexv6",
            "_score": 1,
            "_source": {
              "email": "https://example.com",
              "name": "testnoid"
            },
            "_type": "testtype"
          },
          {
            "_id": "11",
            "_index": "testindexv6",
            "_score": 1,
            "_source": {
              "email": "https://example.com",
              "name": "test11",
              "number": "123"
            },
            "_type": "testtype"
          }
        ],
        "max_score": 1,
        "total": {
          "value": 2
        }
      },
      "shards": {
        "failed": 0,
        "skipped": 0,
        "successful": 1,
        "total": 1
      },
      "timed_out": false,
      "took": 5
    },
    "status": "ok"
  },
  "type": "action_event",
  "version": "v1"
}

```

<summary>
docker run --rm -i rapid7/elasticsearch:3.0.1 --debug run < tests.back/search_documents_v6_without_query.json
</summary>
</details>

<details>

```
{
  "body": {
    "log": "Connect: Warning, No Auth Provided\nrapid7/https://example.com Step name: search_documents\n",
    "meta": {},
    "output": {
      "hits": {
        "hits": [
          {
            "_id": "11",
            "_index": "testindexv6",
            "_score": 0.6931472,
            "_source": {
              "email": "https://example.com",
              "name": "test11",
              "number": "123"
            },
            "_type": "testtype",
            "_version": 2
          }
        ],
        "max_score": 0.6931472,
        "total": {
          "value": 1
        }
      },
      "shards": {
        "failed": 0,
        "skipped": 0,
        "successful": 1,
        "total": 1
      },
      "timed_out": false,
      "took": 13
    },
    "status": "ok"
  },
  "type": "action_event",
  "version": "v1"
}

```

<summary>
docker run --rm -i rapid7/elasticsearch:3.0.1 --debug run < tests.back/search_documents_v6_with_query.json
</summary>
</details>

### Validators
<details>

```
[*] Validating plugin with all validators at .

[*] Running Integration Validators...
[*] Executing validator HelpValidator
[*] Executing validator ChangelogValidator
[*] Executing validator CloudReadyConnectionCredentialTokenValidator
[*] Executing validator RequiredKeysValidator
[*] Executing validator UseCaseValidator
[*] Executing validator SpecPropertiesValidator
[*] Executing validator SpecVersionValidator
[*] Executing validator FilesValidator
[*] Executing validator TagValidator
[*] Executing validator DescriptionValidator
[*] Executing validator TitleValidator
[*] Executing validator VendorValidator
[*] Executing validator DefaultValueValidator
[*] Executing validator IconValidator
[*] Executing validator RequiredValidator
[*] Executing validator VersionValidator
[*] Executing validator DockerfileParentValidator
[*] Executing validator ProfanityValidator
[*] Executing validator AcronymValidator
[*] Executing validator JSONValidator
[*] Executing validator OutputValidator
[*] Executing validator RegenerationValidator
[*] Executing validator HelpInputOutputValidator
[*] Executing validator SupportValidator
[*] Executing validator RuntimeValidator
[*] Executing validator VersionPinValidator
[*] Executing validator EncodingValidator
[*] Executing validator ExampleInputValidator
[*] Executing validator CloudReadyValidator
[*] Executing validator UnapprovedKeywordsValidator
[*] Executing validator HelpExampleValidator
[*] Executing validator ExceptionValidator
[*] Executing validator CredentialsValidator
[*] Executing validator PasswordValidator
[*] Executing validator PrintValidator
[*] Executing validator ConfidentialValidator
[*] Executing validator DockerValidator
[*] Executing validator URLValidator
WARNING: URLs found that return a 4xx code. Verify they are publicly accessible and if not, update with a working URL.
violation: plugin.spec.yaml[30]: https://www.example.com:9243
[*] Plugin failed validation! The following validation errors occurred:

Validator "HelpExampleValidator" failed! 
	Cause: Invalid JSON example identified in Example input of action titled Cluster Health. Please correct the JSON example in help.md.
	Invalid JSON example identified in Example input of action titled Input. Please correct the JSON example in help.md.
	Invalid JSON example identified in Example input of action titled Input. Please correct the JSON example in help.md.
	Invalid JSON example identified in Example input of trigger titled Poll Documents. Please correct the JSON example in help.md.


----
[*] Total time elapsed: 16188.500999999998ms

```

<summary>
icon-validate --all .
</summary>
</details>

<details>

```
[*] Use ``make menu`` for available targets
[*] Including available Makefiles: ../../tools/Makefiles/Colors.mk ../../tools/Makefiles/Helpers.mk
--
[*] Running validators
DEPRECATION: Configuring installation scheme with distutils config files is deprecated and will no longer work in the near future. If you are using a Homebrew or Linuxbrew Python, please see discussion at https://github.com/Homebrew/homebrew-core/issues/76621
[*] Validating plugin at .

[*] Running Integration Validators...
[*] Executing validator HelpValidator
[*] Executing validator ChangelogValidator
[*] Executing validator CloudReadyConnectionCredentialTokenValidator
[*] Executing validator RequiredKeysValidator
[*] Executing validator UseCaseValidator
[*] Executing validator SpecPropertiesValidator
[*] Executing validator SpecVersionValidator
[*] Executing validator FilesValidator
[*] Executing validator TagValidator
[*] Executing validator DescriptionValidator
[*] Executing validator TitleValidator
[*] Executing validator VendorValidator
[*] Executing validator DefaultValueValidator
[*] Executing validator IconValidator
[*] Executing validator RequiredValidator
[*] Executing validator VersionValidator
[*] Executing validator DockerfileParentValidator
[*] Executing validator ProfanityValidator
[*] Executing validator AcronymValidator
[*] Executing validator JSONValidator
[*] Executing validator OutputValidator
[*] Executing validator RegenerationValidator
[*] Executing validator HelpInputOutputValidator
[*] Executing validator SupportValidator
[*] Executing validator RuntimeValidator
[*] Executing validator VersionPinValidator
[*] Executing validator EncodingValidator
[*] Executing validator ExampleInputValidator
[*] Executing validator CloudReadyValidator
[*] Executing validator UnapprovedKeywordsValidator
[*] Executing validator HelpExampleValidator
[*] Plugin failed validation! The following validation errors occurred:

Validator "HelpExampleValidator" failed! 
	Cause: Invalid JSON example identified in Example input of action titled Cluster Health. Please correct the JSON example in help.md.
	Invalid JSON example identified in Example input of action titled Input. Please correct the JSON example in help.md.
	Invalid JSON example identified in Example input of action titled Input. Please correct the JSON example in help.md.
	Invalid JSON example identified in Example input of trigger titled Poll Documents. Please correct the JSON example in help.md.


----
[*] Total time elapsed: 2142.766ms

```

<summary>
make validate
</summary>
</details>

jschipp-r7 commented on plugins/elasticsearch/help.md:
```diff
@@ -257,12 +250,7 @@ Example input:
-    "gender": "M",
-    "city": "Dante"
-  },
+  "document": "{\"firstname\": \"Jon\", \"lastname\": \"Doe\", \"gender\": \"M\", \"city\": \"Dante\"}",
```
Previous example was correctly formatted

r7-mgorny commented:
@jonschipp  About "Also, this plugin supports ElasticSearch 6 and 7. Does the old query work in 6? do we need to check the ES version in the if-statement?. We'll want to make sure it will still work for 6 and 7 users." we change only action. Query which is sending to ES is good in both 7 and 6 version. So this shouldn't be a problem :)

If there are more questions about this maybe we can setup 15 minutes call to discuss it?

r7-mgorny commented:
## Assessment
### Run

<details>

```
{
  "body": {
    "log": "Connect: Connecting..\nrapid7/https://example.com Step name: search_documents\n",
    "meta": {},
    "output": {
      "hits": {
        "hits": [
          {
            "_id": "eB3N2XoBW9uWLGlBkjnw",
            "_index": "testindexv7",
            "_score": 1,
            "_source": {
              "email": "https://example.com",
              "name": "testnoid"
            },
            "_type": "_doc"
          }
        ],
        "max_score": 1,
        "total": {
          "value": 3
        }
      },
      "shards": {
        "failed": 0,
        "skipped": 0,
        "successful": 1,
        "total": 1
      },
      "timed_out": false,
      "took": 2
    },
    "status": "ok"
  },
  "type": "action_event",
  "version": "v1"
}

```

<summary>
docker run --rm -i rapid7/elasticsearch:3.0.2 --debug run < tests/search_documents_v7_without_query.json
</summary>
</details>

<details>

```
{
  "body": {
    "log": "Connect: Connecting..\nrapid7/https://example.com Step name: search_documents\n",
    "meta": {},
    "output": {
      "hits": {
        "hits": [
          {
            "_id": "eB3N2XoBW9uWLGlBkjnw",
            "_index": "testindexv7",
            "_score": 1,
            "_source": {
              "email": "https://example.com",
              "name": "testnoid"
            },
            "_type": "_doc",
            "_version": 1
          }
        ],
        "max_score": 1,
        "total": {
          "value": 3
        }
      },
      "shards": {
        "failed": 0,
        "skipped": 0,
        "successful": 1,
        "total": 1
      },
      "timed_out": false,
      "took": 0
    },
    "status": "ok"
  },
  "type": "action_event",
  "version": "v1"
}

```

<summary>
docker run --rm -i rapid7/elasticsearch:3.0.2 --debug run < tests/search_documents_v7_all_matches_query.json
</summary>
</details>

<details>

```
{
  "body": {
    "error": "An error occurred during plugin execution!\n\nWrong input 'query'. Old query style detected during input. Input shouldn't contain {'query': {'query': ...}}. Please refer help for more details or elasticsearch API documentation: https://example.com",
    "log": "Connect: Connecting..\nrapid7/https://example.com Step name: search_documents\nAn error occurred during plugin execution!\n\nWrong input 'query'. Old query style detected during input. Input shouldn't contain {'query': {'query': ...}}. Please refer help for more details or elasticsearch API documentation: https://example.com\nTraceback (most recent call last):\n  File \"/usr/local/lib/https://example.com\", line 326, in handle_step\n    output = https://example.com\n  File \"/usr/local/lib/https://example.com\", line 476, in start_step\n    output = func(params)\n  File \"/usr/local/lib/https://example.com\", line 22, in run\n    raise PluginException(\https://example.com An error occurred during plugin execution!\n\nWrong input 'query'. Old query style detected during input. Input shouldn't contain {'query': {'query': ...}}. Please refer help for more details or elasticsearch API documentation: https://example.com\n",
    "meta": {},
    "status": "error"
  },
  "type": "action_event",
  "version": "v1"
}

```

<summary>
docker run --rm -i rapid7/elasticsearch:3.0.2 --debug run < tests/search_documents_v7_all_matches_query_bad.json
</summary>
</details>

<details>

```
{
  "body": {
    "log": "Connect: Warning, No Auth Provided\nrapid7/https://example.com Step name: search_documents\n",
    "meta": {},
    "output": {
      "hits": {
        "hits": [
          {
            "_id": "m77N2XoBjtIgvTKiUnVO",
            "_index": "testindexv6",
            "_score": 1,
            "_source": {
              "email": "https://example.com",
              "name": "testnoid"
            },
            "_type": "testtype"
          }
        ],
        "max_score": 1,
        "total": {
          "value": 2
        }
      },
      "shards": {
        "failed": 0,
        "skipped": 0,
        "successful": 1,
        "total": 1
      },
      "timed_out": false,
      "took": 632
    },
    "status": "ok"
  },
  "type": "action_event",
  "version": "v1"
}

```

<summary>
docker run --rm -i rapid7/elasticsearch:3.0.2 --debug run < tests/search_documents_v6_without_query.json
</summary>
</details>

<details>

```
{
  "body": {
    "log": "Connect: Warning, No Auth Provided\nrapid7/https://example.com Step name: search_documents\n",
    "meta": {},
    "output": {
      "hits": {
        "hits": [
          {
            "_id": "11",
            "_index": "testindexv6",
            "_score": 0.6931472,
            "_source": {
              "email": "https://example.com",
              "name": "test11",
              "number": "123"
            },
            "_type": "testtype",
            "_version": 2
          }
        ],
        "max_score": 0.6931472,
        "total": {
          "value": 1
        }
      },
      "shards": {
        "failed": 0,
        "skipped": 0,
        "successful": 1,
        "total": 1
      },
      "timed_out": false,
      "took": 1130
    },
    "status": "ok"
  },
  "type": "action_event",
  "version": "v1"
}

```

<summary>
docker run --rm -i rapid7/elasticsearch:3.0.2 --debug run < tests/search_documents_v6_with_query.json
</summary>
</details>

### Test

Autogenerate with:
<details>

```
{
  "body": {
    "log": "Connect: Connecting..\nrapid7/https://example.com Step name: search_documents\n",
    "meta": {},
    "output": {
      "success": {
        "cluster_name": "r7-cluster",
        "cluster_uuid": "-TK4P6crRBiJpSAqUMUuaQ",
        "name": "elastsearch-xpack",
        "tagline": "You Know, for Search",
        "version": {
          "build_date": "https://example.com",
          "build_flavor": "default",
          "build_hash": "02699626f388ed830012e5b787640e71c56d42d8",
          "build_snapshot": false,
          "build_type": "tar",
          "lucene_version": "https://example.com",
          "minimum_index_compatibility_version": "https://example.com",
          "minimum_wire_compatibility_version": "https://example.com",
          "number": "https://example.com"
        }
      }
    },
    "status": "ok"
  },
  "type": "action_event",
  "version": "v1"
}

```

<summary>
docker run --rm -i rapid7/elasticsearch:3.0.2 --debug test < tests/search_documents_v7_without_query.json
</summary>
</details>

### Test

Autogenerate with:
<details>

```
[*] Use ``make menu`` for available targets
[*] Including available Makefiles: ../../tools/Makefiles/Colors.mk ../../tools/Makefiles/Helpers.mk
--
[*] Running validators
DEPRECATION: Configuring installation scheme with distutils config files is deprecated and will no longer work in the near future. If you are using a Homebrew or Linuxbrew Python, please see discussion at https://github.com/Homebrew/homebrew-core/issues/76621
[*] Validating plugin at .

[*] Running Integration Validators...
[*] Executing validator HelpValidator
[*] Executing validator ChangelogValidator
[*] Executing validator CloudReadyConnectionCredentialTokenValidator
[*] Executing validator RequiredKeysValidator
[*] Executing validator UseCaseValidator
[*] Executing validator SpecPropertiesValidator
[*] Executing validator SpecVersionValidator
[*] Executing validator FilesValidator
[*] Executing validator TagValidator
[*] Executing validator DescriptionValidator
[*] Executing validator TitleValidator
[*] Executing validator VendorValidator
[*] Executing validator DefaultValueValidator
[*] Executing validator IconValidator
[*] Executing validator RequiredValidator
[*] Executing validator VersionValidator
[*] Executing validator DockerfileParentValidator
[*] Executing validator ProfanityValidator
[*] Executing validator AcronymValidator
[*] Executing validator JSONValidator
[*] Executing validator OutputValidator
[*] Executing validator RegenerationValidator
[*] Executing validator HelpInputOutputValidator
[*] Executing validator SupportValidator
[*] Executing validator RuntimeValidator
[*] Executing validator VersionPinValidator
[*] Executing validator EncodingValidator
[*] Executing validator ExampleInputValidator
[*] Executing validator CloudReadyValidator
[*] Executing validator SupportedVersionValidator
[*] Executing validator UnapprovedKeywordsValidator
[*] Executing validator HelpExampleValidator
[*] Executing validator Version Increment Validator
/Users/mgorny/projects/insightconnect-plugins
<git.repo.base.Repo '/Users/mgorny/projects/insightconnect-plugins/.git'>
Traceback (most recent call last):
  File "/usr/local/bin/icon-validate", line 8, in <module>
    sys.exit(main())
  File "/usr/local/lib/python3.9/site-packages/icon_validator/__main__.py", line 52, in main
    return_code = validate(directory=path, spec_file_name=spec_file_name)
  File "/usr/local/lib/python3.9/site-packages/icon_validator/validate.py", line 35, in validate
    v.validate(spec)
  File "/usr/local/lib/python3.9/site-packages/icon_validator/rules/plugin_validators/version_bump_validator.py", line 355, in validate
    remote_spec = VersionBumpValidator.get_remote_spec(spec)
  File "/usr/local/lib/python3.9/site-packages/icon_validator/rules/plugin_validators/version_bump_validator.py", line 54, in get_remote_spec
    blob = VersionBumpValidator.get_plugin_spec_blob(remote_list, spec.spec_dictionary()["name"])
  File "/usr/local/lib/python3.9/site-packages/icon_validator/rules/plugin_validators/version_bump_validator.py", line 76, in get_plugin_spec_blob
    for _blob in _remote.object.tree[RepoConstants.PLUGIN_DIRNAME][plugin_name]:
  File "/usr/local/lib/python3.9/site-packages/gitdb/util.py", line 253, in __getattr__
    self._set_cache_(attr)
  File "/usr/local/lib/python3.9/site-packages/git/objects/commit.py", line 199, in _set_cache_
    self._deserialize(BytesIO(stream.read()))
  File "/usr/local/lib/python3.9/site-packages/git/objects/commit.py", line 528, in _deserialize
    self.tree = Tree(self.repo, hex_to_bin(readline().split()[1]), Tree.tree_id << 12, '')
  File "/usr/local/lib/python3.9/site-packages/git/objects/tree.py", line 215, in __init__
    super(Tree, self).__init__(repo, binsha, mode, path)
  File "/usr/local/lib/python3.9/site-packages/git/objects/base.py", line 168, in __init__
    super(IndexObject, self).__init__(repo, binsha)
  File "/usr/local/lib/python3.9/site-packages/git/objects/base.py", line 56, in __init__
    super(Object, self).__init__()
  File "/usr/local/Cellar/python@3.9/3.9.7_1/Frameworks/Python.framework/Versions/3.9/lib/python3.9/typing.py", line 1083, in _no_init
    raise TypeError('Protocols cannot be instantiated')
TypeError: Protocols cannot be instantiated
make: *** [validate] Error 1

```

<summary>
make validate
</summary>
</details>

<details>

```
[*] Validating plugin with all validators at .

[*] Running Integration Validators...
[*] Executing validator HelpValidator
[*] Executing validator ChangelogValidator
[*] Executing validator CloudReadyConnectionCredentialTokenValidator
[*] Executing validator RequiredKeysValidator
[*] Executing validator UseCaseValidator
[*] Executing validator SpecPropertiesValidator
[*] Executing validator SpecVersionValidator
[*] Executing validator FilesValidator
[*] Executing validator TagValidator
[*] Executing validator DescriptionValidator
[*] Executing validator TitleValidator
[*] Executing validator VendorValidator
[*] Executing validator DefaultValueValidator
[*] Executing validator IconValidator
[*] Executing validator RequiredValidator
[*] Executing validator VersionValidator
[*] Executing validator DockerfileParentValidator
[*] Executing validator ProfanityValidator
[*] Executing validator AcronymValidator
[*] Executing validator JSONValidator
[*] Executing validator OutputValidator
[*] Executing validator RegenerationValidator
[*] Executing validator HelpInputOutputValidator
[*] Executing validator SupportValidator
[*] Executing validator RuntimeValidator
[*] Executing validator VersionPinValidator
[*] Executing validator EncodingValidator
[*] Executing validator ExampleInputValidator
[*] Executing validator CloudReadyValidator
[*] Executing validator SupportedVersionValidator
[*] Executing validator UnapprovedKeywordsValidator
[*] Executing validator HelpExampleValidator
[*] Executing validator Version Increment Validator
/Users/mgorny/projects/insightconnect-plugins
<git.repo.base.Repo '/Users/mgorny/projects/insightconnect-plugins/.git'>
Traceback (most recent call last):
  File "/usr/local/bin/icon-validate", line 8, in <module>
    sys.exit(main())
  File "/usr/local/lib/python3.9/site-packages/icon_validator/__main__.py", line 49, in main
    return_code = validate(directory=path, run_all=True)
  File "/usr/local/lib/python3.9/site-packages/icon_validator/validate.py", line 35, in validate
    v.validate(spec)
  File "/usr/local/lib/python3.9/site-packages/icon_validator/rules/plugin_validators/version_bump_validator.py", line 355, in validate
    remote_spec = VersionBumpValidator.get_remote_spec(spec)
  File "/usr/local/lib/python3.9/site-packages/icon_validator/rules/plugin_validators/version_bump_validator.py", line 54, in get_remote_spec
    blob = VersionBumpValidator.get_plugin_spec_blob(remote_list, spec.spec_dictionary()["name"])
  File "/usr/local/lib/python3.9/site-packages/icon_validator/rules/plugin_validators/version_bump_validator.py", line 76, in get_plugin_spec_blob
    for _blob in _remote.object.tree[RepoConstants.PLUGIN_DIRNAME][plugin_name]:
  File "/usr/local/lib/python3.9/site-packages/gitdb/util.py", line 253, in __getattr__
    self._set_cache_(attr)
  File "/usr/local/lib/python3.9/site-packages/git/objects/commit.py", line 199, in _set_cache_
    self._deserialize(BytesIO(stream.read()))
  File "/usr/local/lib/python3.9/site-packages/git/objects/commit.py", line 528, in _deserialize
    self.tree = Tree(self.repo, hex_to_bin(readline().split()[1]), Tree.tree_id << 12, '')
  File "/usr/local/lib/python3.9/site-packages/git/objects/tree.py", line 215, in __init__
    super(Tree, self).__init__(repo, binsha, mode, path)
  File "/usr/local/lib/python3.9/site-packages/git/objects/base.py", line 168, in __init__
    super(IndexObject, self).__init__(repo, binsha)
  File "/usr/local/lib/python3.9/site-packages/git/objects/base.py", line 56, in __init__
    super(Object, self).__init__()
  File "/usr/local/Cellar/python@3.9/3.9.7_1/Frameworks/Python.framework/Versions/3.9/lib/python3.9/typing.py", line 1083, in _no_init
    raise TypeError('Protocols cannot be instantiated')
TypeError: Protocols cannot be instantiated

```

<summary>
icon-validate --all .
</summary>
</details>

mberezin-r7 commented on plugins/elasticsearch/komand_elasticsearch/actions/search_documents/action.py:
```diff
@@ -15,7 +16,15 @@ def __init__(self):
+
+        if isinstance(query, dict) and query.get("query"):
+            raise PluginException(
+                cause="Wrong input 'query'.",
+                assistance="Old query style detected during input. Input shouldn't contain {'query': {'query': ...}}. "
+                "Please refer help for more details or elasticsearch API documentation: "
+                "https://www.elastic.co/guide/en/elasticsearch/reference/current/query-filter-context.html#query-filter-context-ex",
+            )
```
```suggestion
                cause="Wrong input query format",
                assistance="Old query style detected during input. The input shouldn't contain {'query': {'query': ...}}. "
                "Please refer to the help.md for more details or to the Elasticsearch API documentation: "
                "https://www.elastic.co/guide/en/elasticsearch/reference/current/query-filter-context.html#query-filter-context-ex",
            )
```

mberezin-r7 commented on plugins/elasticsearch/unit_test/test_search_documents.py:
```diff
@@ -121,3 +117,18 @@ def test_search_documents_wrong_object(self, mock_request):
+            )
+
+        self.assertEqual(error.exception.cause, "Wrong input 'query'.")
+        self.assertEqual(
+            error.exception.assistance,
+            "Old query style detected during input. Input shouldn't contain {'query': {'query': ...}}. Please refer "
+            "help for more details or elasticsearch API documentation: https://www.elastic.co/guide/en/elasticsearch/"
+            "reference/current/query-filter-context.html#query-filter-context-ex",
+        )
```
```suggestion
        self.assertEqual(
            error.exception.assistance,
            "Old query style detected during input. The input shouldn't contain {'query': {'query': ...}}. "
                "Please refer to the help.md for more details or to the Elasticsearch API documentation: "
                "https://www.elastic.co/guide/en/elasticsearch/reference/current/query-filter-context.html#query-filter-context-ex",
        )
```

mberezin-r7 replied:
This is a pretty brittle test. We don't necessarily need to test that the error output troubleshooting string is exactly the same. We could simply do `assertRaises(PluginException)`.

mberezin-r7 commented on plugins/elasticsearch/unit_test/test_search_documents.py:
```diff
@@ -121,3 +117,18 @@ def test_search_documents_wrong_object(self, mock_request):
+            )
+
+        self.assertEqual(error.exception.cause, "Wrong input 'query'.")
+        self.assertEqual(
+            error.exception.assistance,
+            "Old query style detected during input. The input shouldn't contain {'query': {'query': ...}}. "
+                "Please refer to the help.md for more details or to the Elasticsearch API documentation: "
+                "https://www.elastic.co/guide/en/elasticsearch/reference/current/query-filter-context.html#query-filter-context-ex",
+        )
```
We don't necessarily need to test that the error output troubleshooting string is exactly the same. We could simply do `assertRaises(PluginException)`.
